test vectors for content keys encode/decode

victor-wei126 opened the pull request:
These are a set of test vectors for content keys encoding/decoding in SSZ for both state and history networks. One concern is that for the history network keys, in the spec, currently `content_type` is still part of the container. However, after some discussion, we decided to take it out of the SSZ container and just prefix it as the first byte when converting a key to bytes. This is why in the test vectors `content_type` is left out of input parameters.

UPDATE:

- [x] Fix ContractStorageTrieProof content-id test vector
- [x] Look into BE vs LE

kdeme reviewed (commented):
lgtm with regards to the state content key test vectors.

Test vectors for history content keys will fail for fluffy currently for the reason that you mention already: They are encoded with prefix (like a SSZ Union) outside of the container.

If we want to change it to the prefixed version, I believe this PR should be accompanied with a spec PR that should get approved/merged first. That would typically be the right order. (Or spec changes done in this PR).

> However, after some discussion, we decided to take it out of the SSZ container and just prefix it as the first byte when converting a key to bytes.

I was not aware of the discussion, so I don't know about the arguments that were made.
Prefixing (SSZ Union) the type is not really necessary as the containers itself are actually of the same type.
That being said, it is allowed according to SSZ spec to have multiple selectors with the same type and it would make it more consistent with the other content keys. (Side note: it does make it slightly more clumsy in the usage in our implementation, but I guess that can be neglected as we already do it for state). Resulting encoded size should remain the same.

I think it would be useful to also add the result of the derivation from `content_key` to `content_id` as test vectors. I think mistakes are more likely to happen there (at least for state). This could be done in a follow-up PR though.

kdeme commented on content-keys-test-vectors.md:
```diff
@@ -0,0 +1,117 @@
+
+This document contains test vectors for content keys to help implementations conform to the specification. 
+
+## Content Keys Encodings
```
Can we have somewhere a split of sections between the state content keys and the history content keys? That would make it more readable/searchable/linkable.

kdeme replied:
And once that is there, links could be added to the test-vectors in the state and history specs

victor-wei126 commented:
The discussion was made here: https://discord.com/channels/890617081744220180/890617129425047583/891031228222930946

Currently, in Trin, I implemented this: https://github.com/ethereum/trin/blob/master/trin-core/src/portalnet/types/content_keys.rs#L54...I assume this is the prefixing method you talked about? I went with this since it was the same way messages were encoded to bytes.

> They are encoded with prefix (like a SSZ Union)

To be honest, I'm still unclear on how the union type works and what exactly is meant by using it. But to make sure I'm understanding--is the prefixing method (what we currently have in trin) the same as using SSZ union? I will definitely look into this more.

kdeme commented:
> assume this is the prefixing method you talked about?

Yes, that looks like it.



> To be honest, I'm still unclear on how the union type works and what exactly is meant by using it. But to make sure I'm understanding--is the prefixing method (what we currently have in trin) the same as using SSZ union?

Putting the content type as prefix before the SSZ serialized type (`Container(....)` in this case) is pretty much the same as serializing an SSZ `Union[Container(...), ...]`. As an `Union[T,...]` serialized is just 1 byte of selector + the serialized `T`. It is not fully the same as there are some additional restrictions (e.g. max selector of 127), see spec https://github.com/ethereum/consensus-specs/blob/dev/ssz/simple-serialize.md#union . 
So practically, the serialized result will be the same, it is just a matter of whether you add that prefix outside of your SSZ implementation or inside it through the use of Union.

This is why we could do this PR (https://github.com/ethereum/portal-network-specs/pull/121/files) without having to change your test vectors.

acolytec3 commented on content-keys-test-vectors.md:
````diff
@@ -0,0 +1,117 @@
+#### Expected Output
+```
+content_key = 0x010f00d1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
+```
````
```suggestion
```content_id = 0xf883705bc68dcd46f9064e8c18dc1c1ae90ba88149b1177bdf1986272779675b
```
```

victor-wei126 replied:
This is a minor organizational issue, but if we put the `content_id` results right underneath the key results, this makes it more concise since we don't have to respecify the inputs and the reader can just glance up to see the inputs. It it not as clean however imo as having a completely separate section for "key to id test vectors" vs "encoding/decoding vectors"

acolytec3 replied:
That's fine.  I don't have a strong opinion either way.  I'll update the content IDs to match now that we've settled on redoing the content-Key as a SSZ union since that impacts the input to the hash and then you can structure how you want.  My tests have them all lumped together as that made the most sense to me.

victor-wei126 replied:
Regarding the `content-id`, I actually got a different one, assuming the hex string you gave above was a result of taking the resulting u256 value, converting it to a byte array, and then hex encoding it. I think it makes a difference whether it is hash(key) vs hash(serialized_key) since the serialized one has offset bytes in it, changing the input. Either way, I got a result of:

hash(raw_key): 29429819488426549164963845897024823308610593491897334245125007239038426786481
hash(serialized_key): 37312724842171933545395799780116690342117774772998923407358080374963771394635

which did not match the U256 you gave:
15025167517633317571792618561170587584740338038067807801482118109695980329625

There is a possible chance I am using a different keccak function as well.

acolytec3 replied:
Javascript doesn't support u256 natively so I'm doing it differently though I'm not sure if that has any impact.  Theoretically converting between bytes, hex strings, and u256 shouldn't matter.  I'm taking the raw bytes array from the serialized key and running that directly through the hash function as here - [SHA256](https://github.com/ethereumjs/ultralight/blob/2eb944d53eafafa45d610d346305b5a100e2bbef/packages/portalnetwork/test/historySubnetwork/types.spec.ts#L11) since our hash function takes a bytes array as input and then converting that to a hex string.  I'll try the SHA256 method from one of the other JS web libraries and see if produces a different result (though it shouldn't if they're implemented the same way).

victor-wei126 replied:
I think maybe making the spec a little clearer might help. I looked back at my code and see that there are a few ways that "hash of key" might be interpreted. It seems that `H(key)` vs `H(serialized_key)` gives the same output, but when we say hash of serialized key we might want to specify `H(hexEncode(serialized_key))` because what I had interpreted as hashing the key originally was the raw concatenation of the fields of the key, ie, just concatenating the raw bytes and using that as input w/o any hex encoding.

Having made the key hex encoded, I had the same input string as you had 0x000f00... but still got a different output. I have been using [keccak](https://docs.rs/sha3/0.10.0/sha3/) hash function, from the sha3 family, not sha-256 from the sha2 family, which I believe is what you are using. Maybe try https://www.npmjs.com/package/sha3?

Should we change the spec to use keccak256?

acolytec3 replied:
The history network spec says specifically to use SHA-256 to derive the `Content-ID` - https://github.com/ethereum/portal-network-specs/blob/master/history-network.md#content-id

I don't have an opinion about sha256 vs keccak but it would seem that at least an initial issue your code isn't following the current spec.

acolytec3 replied:
But, I'm happy to switch if it makes sense to do so.

acolytec3 commented on content-keys-test-vectors.md:
````diff
@@ -0,0 +1,117 @@
+#### Expected Output
+```
+content_key = 0x021400d1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
+```
````
```suggestion
```content_id = 0xfccf5551a4b704080d4a7c89fc87b3e88a232683c88c2100240e05b6cf2fe441
```

acolytec3 commented on content-keys-test-vectors.md:
````diff
@@ -0,0 +1,117 @@
+#### Expected Output
+```
+content_key = 0x030400d1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
+```
````
```suggestion
```content_id = 0x42a3a15338bbd23dd373f831eddbb48d781120c2ea052cc06ff839875e5682e9
```

acolytec3 reviewed (commented):
I've added the content IDs produced by the content_keys (assuming the content key is `content_type + sszContainer(chain_id, block_hash)` (though this is the same as the SSZ Union type we're using in Ultralight)

acolytec3 commented on content-keys-test-vectors.md:
````diff
@@ -0,0 +1,121 @@
+
+##### Expected Output
+```
+content_key = 0x010f00d1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
````
```suggestion
content_key = 0x000f00d1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
```

acolytec3 commented on content-keys-test-vectors.md:
````diff
@@ -0,0 +1,121 @@
+
+##### Expected Output
+```
+content_key = 0x021400d1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
````
```suggestion
content_key = 0x011400d1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
```

acolytec3 commented on content-keys-test-vectors.md:
````diff
@@ -0,0 +1,121 @@
+
+##### Expected Output
+```
+content_key = 0x030400d1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
````
```suggestion
content_key = 0x020400d1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
```

acolytec3 reviewed (commented):
I've updated the expected resulting vectors to reflect the new `content-type` numbering and the content IDs to match based on the recently merged PR #124.

victor-wei126 commented:
I added the `content-id` test vector for state network keys. @acolytec3 let me know if you may need the hex string instead of the u256 value since js doesn't support it. Also for these derivations, I am also using keccak, and not sha256, as the spec currently says. So either we update the state network spec or I can change mine to use sha256. What are your guys thoughts on this, is there any particular difference in choosing one or the other?

kdeme commented:
> I've updated the expected resulting vectors to reflect the new content-type numbering and the content IDs to match based on the recently merged PR #124.

I get the same results now for all the content key serializations (after the Union change).

For the content ids in History, I get the same results as the byte arrays @acolytec3 posted (I also have the resulting uint256 as BE from those arrays available).

For the content ids in state, they don't match with what is currently in this PR, but I'll assume they are using still keccak everywhere?

victor-wei126 commented:
Yep, I've just updated the content ids for state, using sha256 instead of keccak. I added both the U256 and hex string result. One thing to note: for `ContractStorageTrieProofKey`, when taking the slot hash, I am currently converting the slot (a u256 value) to a string, then hex decoding that string into a byte array, which is passed to sha256. I guess the correct way might be to directly convert the slot u256 value into a byte array in LE format. How are you guys implementing this?

I got the same results for history as @acolytec3 as well now so the history keys should be good.

acolytec3 reviewed (approved):
LGTM as far as the History Network vectors go.  As noted on Discord, I haven't implemented the State Network encoding pieces yet so can't comment there.

kdeme reviewed (approved):
After last updates, all content and history keys/ids match now with implementation in Fluffy.

Should note though that Endianness is currently under-specified. Current test vectors go with BE for hash -> id (uint256), and also with BE for slot (uint256) -> array for hashing

victor-wei126 commented:
I could use some clarification with the endianness issue.

> Current test vectors go with BE for hash -> id (uint256)

I assume from this you mean the sha256 hash to the uint256 conversion. The hash function outputs a 32 byte array, and I assume this is in BE since that is "natural" order. So when we convert to u256, wouldn't we always want BE, as LE would give us the wrong U256? So I'm confused why endianness is an issue here.

> BE for slot (uint256) -> array for hashing

By this you mean whether we are taking the slot value as LE or BE? Again, it seems to me BE is the only natural choice since that gives us the "correct" u256 value.

pipermerriam commented on content-keys-test-vectors.md:
````diff
@@ -0,0 +1,142 @@
+```
+content_key = 0x020400d1c390624d3bd4e409a61a858e5dcc5517729a9170d014a6c96530d64dd8621d
+content_id = 0xaa39e1423e92f5a667ace5b79c2c98adbfd79c055d891d0b9c49c40f816563b2
+content_id: U256 = 76995449220721979583200368506411933662679656077191192504502358532083948020658
````
What is the reason for presenting this as both an integer and the hex value?

This de-normalization seems a little un-necessary and has the potential downside of the two values not actually being the same due to clerical errors.

victor-wei126 replied:
The reason is because U256 is the actual type of the content-id so it'd be nice to have it there but not all languages support it natively, like JS, so we have the hex value as well.
